Miscellaneous updates and bug fixes

wesleykendall opened the pull request.

PaulGilmartin commented on pgtransaction/tests/test_atomic.py:
```diff
@@ -88,77 +83,58 @@ class MockError(Exception):
+            with atomic():  # pragma: no branch
+                trade.id = None
+                trade.save()
+        except IntegrityError:
```
Any particular reason we switched to IntegrityError here? Main idea of this test was to demonstrate that any exception, not just ones coming from db operations, behaved as expected with respect to nested atomics (i.e. savepoints).

wesleykendall replied:
I see, will change back!

wesleykendall replied:
Addressed in 4acadc0e7f3a6f368444a63c2a0187b6370c0750

PaulGilmartin commented on pgtransaction/atomic.py:
```diff
@@ -19,32 +18,36 @@ def __init__(
-                "inherit the isolation level from their parent transaction "
-                "automatically."
-            )
+        if self.isolation_level:
```
This should probably be `if self.isolation_level or self.retry`, since the retry policy only works currently for psycopg2 errors.

wesleykendall replied:
I'd argue against checking that `retry` is set. Although we currently only configure it to only catch psycopg2 errors, the code will still work if you switch out the DB as long as psycopg2 is installed. It just won't actually retry anything.

For example, say that a user uses sqlite for their test suite and only uses `retry` and never `isolation_level`. Although this might be a rare edge case, I like to avoid doing things that might be overly restrictive.

PaulGilmartin replied:
Fair point about the tests. I suppose if we make it clear we only catch postgres errors for now then it should be safe enough.

PaulGilmartin commented on pgtransaction/atomic.py:
```diff
@@ -18,17 +18,18 @@ def __init__(
         self._validate()
 
     def _validate(self):
+        connection = transaction.get_connection(self.using)
```
Great catch, of course when this is a decorator all of this happens at module load time.

wesleykendall replied:
Yea it totally crossed my mind too until I saw a failure in a test. Thank goodness for the test suite!

PaulGilmartin commented on pgtransaction/tests/test_atomic.py:
```diff
@@ -113,103 +114,148 @@ def test_atomic_retries_context_manager_not_allowed():
+
+    t1.join()
+    t2.join()
+
```
Really cool you got this to work. I tried pretty much exactly this set up minus the "barrier" thing (which I've never even heard of) and could not get things to work. I wonder barrier was the missing piece.

wesleykendall replied:
Yea the barrier plus the small sleep time makes it pretty reliable

PaulGilmartin commented on pgtransaction/atomic.py:
```diff
@@ -58,38 +59,40 @@ def __call__(self, func):
-                    else:
+                        or num_retries >= self.retry
+                    ):
                         raise
```
If we raise a SerializationFailure or DeadlockDetected error here, aren't we leaving the data from the transaction where those errors were invoked in an unfinished state? What's rolling back the data correctly if we don't manually invoke __exit__ as was done previously?

wesleykendall replied:
`with self._recreate_cm()` is inside of the try/except, so the transaction is rolled back before we continue the retry loop

PaulGilmartin replied:
Ahh yes of course. This way makes much more sense than the original.

wesleykendall commented:
@PaulGilmartin addressed comments and put a RuntimeError back in that test you mentioned.

I left two convos unresolved if you have any additional thoughts in case I might be overlooking something. It would be nice if you can double check my logic for the transaction closing. I believe it's fine, and the multi-threaded test should also properly ensure that the transaction is being reverted when retrying
